Indian Districts for COVID19

edumorlom opened the pull request:
Import for Indian Districts and States from covid19india.org.
Each state has its own API.
Place names are resolved by wikidataId.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/datacommonsorg/data/328)
<!-- Reviewable:end -->

tjann commented on scripts/covid19indiaORG/INDIA_MAP.py:
```diff
@@ -0,0 +1,722 @@
+   "Arunachal Pradesh":"Q1162",
+   "Andhra Pradesh":"Q1159",
+   "Andaman and Nicobar Islands":"Q40888"
+}
```
How did you come up with this map?

edumorlom replied:
The places' dcids are resolved by wikidataId.
The map of State->District->wikidataId was generated using the following:
1. The wikidataId for each place is queried using the place_resolver.go script.
2. A script is used against wikidata.org/wiki/${wikidataId} that verifies that the place is both a District and part of India.
3. Manual check has been performed to ensure that the name matches.

tjann replied:
1. I see. Does the data have any better IDs?
2. Do you have that code too? Should we check it in?
3. How many places did you manually check? Ballpark is fine.

tjann replied:
@pradh do you have any opinions about manifesting this import to prod if we are using place name resolver to get the Wikidata IDs of Indian districts?

edumorlom replied:
@tjann I have added the script that checks against wikidataId to ensure the wikidataId is correct.
It basically goes through all the wikidataIds and exports a CSV with wikidata name and whether it belongs to India.
Then it's very easy to manually check that all are correct. I added a README.md too.

pradh replied:
This dataset does not have any IDs right?  So approach to mapping to wikidataId via some heuristics (including place name resolver) and manual check, sounds reasonable...

tjann commented on scripts/covid19indiaORG/Config.py:
```diff
@@ -0,0 +1,177 @@
+   "State": "Andaman and Nicobar Islands",
+   "API": "https://api.covid19india.org/v4/min/timeseries-AN.min.json"
+ }
+]
```
Is this autogenerated or manually done?

edumorlom replied:
I autogenerated this from the following:

`<select><option value="{&quot;stateCode&quot;:&quot;TT&quot;,&quot;districtName&quot;:null}">India</option><option value="{&quot;stateCode&quot;:&quot;AN&quot;,&quot;districtName&quot;:null}">Andaman and Nicobar Islands</option><option value="{&quot;stateCode&quot;:&quot;AP&quot;,&quot;districtName&quot;:null}">Andhra Pradesh</option><option value="{&quot;stateCode&quot;:&quot;AR&quot;,&quot;districtName&quot;:null}">Arunachal Pradesh</option><option value="{&quot;stateCode&quot;:&quot;AS&quot;,&quot;districtName&quot;:null}">Assam</option><option value="{&quot;stateCode&quot;:&quot;BR&quot;,&quot;districtName&quot;:null}">Bihar</option><option value="{&quot;stateCode&quot;:&quot;CH&quot;,&quot;districtName&quot;:null}">Chandigarh</option><option value="{&quot;stateCode&quot;:&quot;CT&quot;,&quot;districtName&quot;:null}">Chhattisgarh</option><option value="{&quot;stateCode&quot;:&quot;DN&quot;,&quot;districtName&quot;:null}">Dadra and Nagar Haveli and Daman and Diu</option><option value="{&quot;stateCode&quot;:&quot;DL&quot;,&quot;districtName&quot;:null}">Delhi</option><option value="{&quot;stateCode&quot;:&quot;GA&quot;,&quot;districtName&quot;:null}">Goa</option><option value="{&quot;stateCode&quot;:&quot;GJ&quot;,&quot;districtName&quot;:null}">Gujarat</option><option value="{&quot;stateCode&quot;:&quot;HR&quot;,&quot;districtName&quot;:null}">Haryana</option><option value="{&quot;stateCode&quot;:&quot;HP&quot;,&quot;districtName&quot;:null}">Himachal Pradesh</option><option value="{&quot;stateCode&quot;:&quot;JK&quot;,&quot;districtName&quot;:null}">Jammu and Kashmir</option><option value="{&quot;stateCode&quot;:&quot;JH&quot;,&quot;districtName&quot;:null}">Jharkhand</option><option value="{&quot;stateCode&quot;:&quot;KA&quot;,&quot;districtName&quot;:null}">Karnataka</option><option value="{&quot;stateCode&quot;:&quot;KL&quot;,&quot;districtName&quot;:null}">Kerala</option><option value="{&quot;stateCode&quot;:&quot;LA&quot;,&quot;districtName&quot;:null}">Ladakh</option><option value="{&quot;stateCode&quot;:&quot;MP&quot;,&quot;districtName&quot;:null}">Madhya Pradesh</option><option value="{&quot;stateCode&quot;:&quot;MH&quot;,&quot;districtName&quot;:null}">Maharashtra</option><option value="{&quot;stateCode&quot;:&quot;MN&quot;,&quot;districtName&quot;:null}">Manipur</option><option value="{&quot;stateCode&quot;:&quot;ML&quot;,&quot;districtName&quot;:null}">Meghalaya</option><option value="{&quot;stateCode&quot;:&quot;MZ&quot;,&quot;districtName&quot;:null}">Mizoram</option><option value="{&quot;stateCode&quot;:&quot;NL&quot;,&quot;districtName&quot;:null}">Nagaland</option><option value="{&quot;stateCode&quot;:&quot;OR&quot;,&quot;districtName&quot;:null}">Odisha</option><option value="{&quot;stateCode&quot;:&quot;PY&quot;,&quot;districtName&quot;:null}">Puducherry</option><option value="{&quot;stateCode&quot;:&quot;PB&quot;,&quot;districtName&quot;:null}">Punjab</option><option value="{&quot;stateCode&quot;:&quot;RJ&quot;,&quot;districtName&quot;:null}">Rajasthan</option><option value="{&quot;stateCode&quot;:&quot;SK&quot;,&quot;districtName&quot;:null}">Sikkim</option><option value="{&quot;stateCode&quot;:&quot;TN&quot;,&quot;districtName&quot;:null}">Tamil Nadu</option><option value="{&quot;stateCode&quot;:&quot;TG&quot;,&quot;districtName&quot;:null}">Telangana</option><option value="{&quot;stateCode&quot;:&quot;TR&quot;,&quot;districtName&quot;:null}">Tripura</option><option value="{&quot;stateCode&quot;:&quot;UN&quot;,&quot;districtName&quot;:null}">Unassigned</option><option value="{&quot;stateCode&quot;:&quot;UP&quot;,&quot;districtName&quot;:null}">Uttar Pradesh</option><option value="{&quot;stateCode&quot;:&quot;UT&quot;,&quot;districtName&quot;:null}">Uttarakhand</option><option value="{&quot;stateCode&quot;:&quot;WB&quot;,&quot;districtName&quot;:null}">West Bengal</option></select>`

tjann replied:
Ah. Do you have the code for this? If it's just a bash script you can even just add to README.

edumorlom replied:
done

tjann commented on scripts/covid19indiaORG/README.md:
```diff
@@ -0,0 +1,63 @@
+### Overview
+
+This dataset shows how Covid-19 has impacted States and Districts in India.
+#### Active Cases
```
Does cumulative active = simple count of currently active?

edumorlom replied:
Sorry, didn't mean to put cumulative in front of active.
Fixed.

tjann commented on scripts/covid19indiaORG/README.md:
```diff
@@ -0,0 +1,63 @@
+
+## Import Artifacts
+
+### Data APIs
+- [Config.py](Config.py)
+
+### India wikidataId Map
+- [INDIA_MAP.py](INDIA_MAP.py)
```
Can you add notes on how you generated these two files?

edumorlom replied:
Done

tjann commented on scripts/covid19indiaORG/README.md:
```diff
@@ -0,0 +1,63 @@
+- [INDIA_MAP.py](INDIA_MAP.py)
+
+### Output CSV
+- [output.mcf](output/output.csv)
```
csv? the text says mcf

edumorlom replied:
Fixed.

tjann commented on scripts/covid19indiaORG/README.md:
```diff
@@ -0,0 +1,63 @@
+- [output.mcf](output/output.csv)
+
+### Scripts
+- [main.py](main.py)
```
short description of what this does?

edumorlom replied:
done

tjann commented on scripts/covid19indiaORG/README.md:
```diff
@@ -0,0 +1,63 @@
+### Scripts
+- [main.py](main.py)
+
+### Notes
+The places' dcids are resolved by wikidataId.
+The wikidataId for each place was found using the place_resolver.go as well as so manual checks.
```
I think this corresponds to the INDIA_MAP.py file? Can you move the text here up? Also, "so manual checks" -- can you clarify how thoroughly you checked it?

edumorlom replied:
Fixed

edumorlom replied:
I added more detailed description.

I basically compared against wikidata.org/wiki/${wikidataId} to make sure that the wikidataId corresponds to a district and to India.
Then I manually checked that wikidataId matches the same name.

tjann commented on scripts/covid19indiaORG/Config.py:
```diff
@@ -0,0 +1,177 @@
+state_apis = [
+ {
+   "Abbreviation": "WB",
+   "State": "West Bengal",
```
Do we need the "State" field?

Is the API always the same? `https://api.covid19india.org/v4/min/timeseries-<Abbreviation>.min.json` ? I am wondering if we can just use a list of Abbreviation.

edumorlom replied:
The reason for the State field is because the DISTRICTS map is keyed by State->District.
The API the District Names, I need the name of the State.

Although, another workaround is to key the DISTRICTS map by abbreviation->district.

tjann replied:
I see. So Abbreviation is not unique within India? I was hoping the values correspond to the isoCodes https://screenshot.googleplex.com/6dSiQeFsXniXqu3

tjann replied:
This might be a stretch, but IF these are isoCodes, it would be nice to give isoCodes for states, since there is more certainty. For districts, if there are no codes, we can continue with your name to wikidata approach.

edumorlom replied:
Upon looking, the abbreviation is actually the iso code minus the IN- in front.
The iso code is IN-WB but they use WB.

I have changed the INDIA hashmaps to be referenced by iso code instead of the name.

So now, they are hashed by: stateIsoCode -> district_name -> wikidataId.

edumorlom replied:
<img width="683" alt="Screen Shot 2020-12-02 at 12 44 08 PM" src="https://user-images.githubusercontent.com/11444894/100910498-13671f80-349c-11eb-8021-2f8416f7a1c5.png">

edumorlom replied:
Should I add the IN- infront? Their APIs don't use the IN-.

tjann replied:
no that's fine, I was thinking we could resolve with isoCode, but if you manually mapped iso<>wikidataId, then it should be good.

tjann reviewed (commented):
I'll review the script after getting a better idea from the current comments!

tjann commented on scripts/covid19indiaORG/main.py:
```diff
@@ -0,0 +1,122 @@
+def _parse_timeseries(json: Dict[str, Dict[str, int]]) -> Dict[str, Dict[str, int]]:
+    """
+    Given Covid 19 India JSON data for a specific place, return the total data
+    in the form of {date: {type_of_data: value}}
```
date or dates?

tjann replied:
Nvm I see.

tjann commented on scripts/covid19indiaORG/main.py:
```diff
@@ -0,0 +1,122 @@
+
+    # Store the DataFrame for this state under the main table.
+    table = pd.concat([table, state_table])
+
+    # The district data is stored under 'districts'.
+    if 'districts' not  in state_data:
+        continue
+    districts_data = state_data['districts']
+
+    # For every district in the state, parse the timeseries.
+    for district_name, data in districts_data.items():
+        district_timeseries = _parse_timeseries(data)
+        # Store the data in a DataFrame, and transpose it.
+        # Transposing puts the dates in the Y-axis.
+        district_table = pd.DataFrame(district_timeseries).T
+
+        if not (state_name in DISTRICTS and district_name in DISTRICTS[state_name]):
+            continue
+
+        # Get the WikidataId from the dictionary.
+        district_table['wikidataId'] = DISTRICTS[state_name][district_name]
+
+        # Store the DataFrame for this district under the main table.
+        table = pd.concat([table, district_table])
+
+# Rename the Y-axis to be "date".
+table.index.name = "date"
+
+# Calculate the active cases = confirmed cases - recovered cases.
+table['active'] = table['confirmed'] - table['recovered']
+
+# Get rid of any rows gthat don't have a wikidataId.
+table.dropna(subset=['wikidataId'], inplace=True)
+
+# Get rid of any rows that don't have at least one of the following.
+table.dropna(subset=['confirmed', 'deceased', 'tested', 'active', 'recovered'], thresh=1, inplace=True)
+
+# Only keep the following columns, the rest are not part of the import.
+table = table[['confirmed', 'deceased', 'tested', 'active', 'recovered', 'wikidataId']]
+
```
Can you add an integration test for this main section? You'll probably want to put this whole chunk into a function called in main. Example integration test: https://github.com/datacommonsorg/data/blob/master/scripts/us_bea/states_gdp/import_data_test.py#L66

edumorlom replied:
Done

edumorlom commented:
Ready for re-review!
1. Added 3 unit tests, instead of API calls, it reads JSON from the file.
2. Minor modifications to code to allow testing.
3. Modifications to README.md.
4. I no longer use the State name, only State isoCode and district name.

tjann commented on scripts/covid19indiaORG/Covid19IndiaORG.mcf:
```diff
@@ -0,0 +1,7 @@
+statType: dcs:measuredValue
+measuredProperty: dcs:count
+medicalStatus: dcs:ConfirmedCase
+medicalCondition: dcs:COVID_19
```
I am thinking we should stay consistent with https://datacommons.org/browser/CumulativeCount_MedicalConditionIncident_COVID_19_ConfirmedCase

and have incidentType instead of medicalCondition

edumorlom replied:
done

tjann commented on scripts/covid19indiaORG/README.md:
```diff
@@ -0,0 +1,104 @@
+# Importing covid19india.org Data Into Data Commons
+
+Author: @edumorales
```
You can use your github handle here

edumorlom replied:
done!

tjann commented on scripts/covid19indiaORG/README.md:
```diff
@@ -0,0 +1,104 @@
+# Importing covid19india.org Data Into Data Commons
```
IIUC, this is only district level, even though the dataset includes state? If that's right, can you make that clear somewhere here? Thanks!

edumorlom replied:
This import includes both district and state data. I will make a note. Thanks!

tjann commented on scripts/covid19indiaORG/import-config.txt:
```diff
@@ -0,0 +1,11 @@
+import: {
+  import_name: "covid19indiaORG"
```
Covid19India

edumorlom replied:
I used covid19indiaORG because there was some other covid19 India import, didn't want to confuse them. But, done!

tjann commented on scripts/covid19indiaORG/import-config.txt:
```diff
@@ -0,0 +1,11 @@
+import: {
+  import_name: "covid19indiaORG"
+  provenance_url: "https://www.covid19india.org/"
+  provenance_description: "Indian State and District Covid19 data."
```
Just district, I thought?

edumorlom replied:
Import includes State data as well.

tjann commented on scripts/covid19indiaORG/import-config.txt:
```diff
@@ -0,0 +1,11 @@
+  import_name: "covid19indiaORG"
+  provenance_url: "https://www.covid19india.org/"
+  provenance_description: "Indian State and District Covid19 data."
+  # Path to the resolved MCF. Can be a repeated field.
```
Remove comment?

edumorlom replied:
done!

tjann commented on scripts/covid19indiaORG/run_tests.py:
```diff
@@ -0,0 +1,79 @@
+        @returns assertion that output.csv == expected.csv.
+        """
+
+        output_path: str = path.join(dir_path, "output.csv")
```
FWIW I think you can use  io.StringIO() for the output. Here is some example:

```
    f_in = gfile.Open(csv_file, 'rt')
    f_out = io.StringIO()
    Covid19IndiaORG(f_in, f_out)
    self.assertEqual(f_out.getvalue(), expected_mcf)
```

Then there's no need to save and update the output file.

edumorlom replied:
Done, thanks!

edumorlom replied:
done

tjann commented on scripts/covid19indiaORG/run_tests.py:
```diff
@@ -0,0 +1,79 @@
+        if not path.exists(output_path):
+            self.fail(output_path + " doesn't exist!")
+
+        # Read the CSV file and generate a DataFrame with it.
+        actual_df = pd.read_csv(output_path)
+        expected_df = pd.read_csv(expected_path)
+
+        # Assert that both dataframes are equal, regardless of order and dtype.
+        pd.testing.assert_frame_equal(
+            actual_df.sort_index(axis=1),
+            expected_df.sort_index(axis=1),
+            check_dtype=False)
```
Can just do string checking instead of reading into pd dataframe

edumorlom replied:
I can do that but then I have to ensure the column order is the same.

tjann replied:
Oh I see. This is a result of using pd df's in the library? We can keep this then

edumorlom replied:
Yeah it is but I changed it, now it always exports the columns by alphabetical order. So CSV files should be identical.

tjann commented on scripts/covid19indiaORG/Covid19IndiaORG.py:
```diff
@@ -0,0 +1,167 @@
+
+    return state_table
+
+def Covid19IndiaORG(state_to_source: Dict[str, str] = None, output: str = "output.csv"):
```
What if we didn't have default values, and instead passed in STATE_APIS and "output.csv" from 'main'  ? I feel that is cleaner and less code.

edumorlom replied:
done.

tjann commented on scripts/covid19indiaORG/Covid19IndiaORG.py:
```diff
@@ -0,0 +1,167 @@
+        downloaded_data: Dict[str, Dict] = _download_data(data_source)
+
+        # If there is no wikidataId for the state, skip it.
+        if iso_code not in STATES:
```
Can you tell me about how much this happens? Maybe also leave a note in README

edumorlom replied:
it doesn't really happen at all, it's just an edge case in case they add some other form of state and we don't have that state in the hashmap.

edumorlom commented:
Ready for re-review!
1. I added the script that checks against wikidata to ensure that the wikidataIds are correct, contains README.md
2. Removed default parameters.
3. Tests will use StringIO.
4. Minor changes to code and documentation.

Thanks

shifucun commented on scripts/covid19indiaORG/CheckWikidataId/CheckWikidataId.py:
```diff
@@ -0,0 +1,52 @@
+# INDIA_MAP is outside the directory, so we have to change the PYTHONPATH.
```
Add license header?

edumorlom replied:
Added, thanks

shifucun commented on scripts/covid19indiaORG/Config.py:
```diff
@@ -0,0 +1,56 @@
+DATA_TO_KEEP = set(['confirmed', 'deceased', 'tested', 'active', 'recovered'])
+
+# Every state has its own API.
+STATE_APIS = {
```
could this just be a function from, say, "WB" to "https://api.covid19india.org/v4/min/timeseries-WB.min.json" if the format of the string is the same

edumorlom replied:
Done!
